_posts/2016-05-12-publish-subscribe-model-in-kafka.md

```diff
@@ -0,0 +1,123 @@
+---
+layout: post
+name: Publish-subscribe-model-in-Kafka
+title: 'Publish-subscribe model in Kafka'
+date: 2016-05-12 00:20:00 +00:00
+author: Felipe Fernández
+image:
+    src: /assets/img/custom/blog/kafka.jpg
+    attribution:
+      text: Kafka Museum by Priit Tammets
+      href: https://www.flickr.com/photos/tammets/2116102744
+tags:
+- publish-subscribe
+- kafka
+- messaging
+
+---
+
+This is the fourth post in the series about integrating sync clients with async systems ([1, ](http://codurance.com/2016/04/28/async-systems-with-sync-clients/)[2, ](http://codurance.com/2016/04/30/akka-basics/)[3](http://codurance.com/2016/05/10/finite-state-machines-with-akka/)). Here we'll try to understand how [Kafka](http://kafka.apache.org/) works in order to leverage properly its publish-subscribe implementation.
+
+
+## Kafka concepts
+
+According to [official documentation](http://kafka.apache.org/documentation.html#introduction):
+
+> Kafka is a distributed, partitioned, replicated commit log service. It provides the functionality of a messaging system, but with a unique design.
+
+Kafka runs as a cluster and the nodes are called brokers. Brokers can be leaders or replicas to provide high-availability and fault tolerance. Brokers are in charge of partitions, being the distribution unit where messages are stored. Those messages are ordered and they're accessible by offset. A set of partitions forms a topic, being a feed of messages. A partition can have different consumers, and they access to the messages using its own offset. Producers publish messages into Kafka topics.
+
+## Queuing vs publish-subscribe
+
+Consumer groups is a key concept to explain why Kafka is more flexible and powerful than other messaging solutions like [RabbitMQ](https://www.rabbitmq.com/). Consumers are associated to consumer groups. If every consumer has the same consumer group, the topic's messages will be evenly load balanced between consumers; that's called queuing model. By contrast, if every consumer has a different consumer group, all the messages will be consumed in every client; that's called publish-subscribe model.
+
+You can have a mix of both approaches, having different logical consumer groups, for different needs, and several consumers inside of every group to increase throughput through parallelism.
+
+## Understanding our needs
+
+<img src="/assets/img/custom/blog/law_enforcement.png" alt="Law enforcement architecture" title="Law enforcement architecture" class="img img-center img-responsive style-screengrab">
+
+As we saw in previous posts the Items service publish messages into a Kafka topic called `item_deleted`. To define in which topic's partition will live that message Kafka provides [three alternatives](https://kafka.apache.org/090/javadoc/org/apache/kafka/clients/producer/Partitioner.html):
```

Comment:
To define in which topic's partition will live that message Kafka provides ==> are you trying to say 

"This is done to define the topic's partition the message provided by Kafka will live" 

or 

"This is done to define the topic's partition the message provided by Kafka will be placed"

or 

"This is done to define the topic's partition the message provided by Kafka will find its place"